app/src/main/java/com/github/se/eventradar/viewmodel/qrCode/QrCodeFriendViewModel.kt

```diff
@@ -0,0 +1,116 @@
+package com.github.se.eventradar.viewmodel.qrCode
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.github.se.eventradar.model.Resource
+import com.github.se.eventradar.model.User
+import com.github.se.eventradar.model.repository.user.IUserRepository
+import com.google.firebase.auth.FirebaseAuth
+import dagger.hilt.android.lifecycle.HiltViewModel
+import javax.inject.Inject
+import kotlinx.coroutines.async
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.filterNotNull
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withTimeoutOrNull
+
+enum class NavigationEvent {
+  None,
+  NavigateToNextScreen
+}
+// ViewModel & UI can be improved on by having a state where the UI reflects a loading icon if
+// firebase operations take a long time
+// would also need to handle the case where the updates really do fail within 1 minute's time.
+// personally gonna make this extra.
+
+@HiltViewModel
+class QrCodeFriendViewModel
+@Inject
+constructor(
+    private val firebaseUserRepository: IUserRepository, // Dependency injection
+    qrCodeAnalyser: QrCodeAnalyser // Dependency injection
+) : ViewModel() {
+
+  private val _decodedResult = MutableStateFlow<String?>(null)
+
+  private val _navigationEvent = MutableStateFlow(NavigationEvent.None)
+  val navigationEvent: StateFlow<NavigationEvent> = _navigationEvent.asStateFlow()
+
+  init {
+    qrCodeAnalyser.onDecoded = { decodedString ->
+      _decodedResult.value = decodedString ?: "Failed to decode QR Code"
+    }
+
+    viewModelScope.launch {
+      _decodedResult
+          .filterNotNull() // Only continue if the value is not null
+          .collect { result ->
+            if (result != "Failed to decode QR Code") {
+              updateFriendList(result)
+            }
+          }
+    }
+  }
+
+  private fun updateFriendList(friendID: String) {
+    viewModelScope.launch {
+      val myUID = FirebaseAuth.getInstance().currentUser!!.uid
+
+      val friendUserDeferred = async { firebaseUserRepository.getUser(friendID) }
```

Comment:
I believe this could all go into one `runBlocking` block (but needs to be tested)